dashboard/app/models/user_level.rb

```diff
@@ -108,6 +110,21 @@ def paired?
     driver? || navigator?
   end
 
+  def csf_validated?
+    Script.get_from_cache(script_id).csf? && Script.cache_find_level(level_id).validated?
+  end
+
+  # Record the time spent on a level by creating or updating the ValidatedUserLevel table
+  # for the UserLevel
+  def record_time_spent(time)
+    return unless time > 0 && csf_validated?
+    Retryable.retryable on: [Mysql2::Error, ActiveRecord::RecordNotUnique], matching: /Duplicate entry/ do
+      vul = ValidatedUserLevel.find_or_initialize_by(user_level_id: id)
```

Comment:
With the `has_one :validated_user_level` [association](https://guides.rubyonrails.org/association_basics.html#has-one-association-reference) still on the UserLevel model, this could be rewritten as:

```suggestion
      vul = validated_user_level || build_validated_user_level
```

Otherwise, the association isn't used and could be removed.